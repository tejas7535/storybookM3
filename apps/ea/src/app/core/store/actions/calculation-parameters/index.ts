export * as CalculationParametersActions from './calculation-parameters.actions';
export * as CalculationTypesActions from './calculation-types.actions';
