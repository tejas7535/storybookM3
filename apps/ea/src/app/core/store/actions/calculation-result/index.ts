export * as CatalogCalculationResultActions from './catalog-calculation-result.actions';
export * as CO2UpstreamCalculationResultActions from './co2-upstream-calculation-result.actions';
export * as CO2DownstreamCalculationActions from './downnstream-calculation.actions';
