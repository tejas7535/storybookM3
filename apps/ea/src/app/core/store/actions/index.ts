export * from './calculation-parameters';
export * from './calculation-result';
export * from './product-selection';
export * from './settings';
export * from './storage-messages';
