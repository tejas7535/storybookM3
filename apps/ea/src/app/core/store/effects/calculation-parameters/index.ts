export * from './calculation-parameters.effects';
