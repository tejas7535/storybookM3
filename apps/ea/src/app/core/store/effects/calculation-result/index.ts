export * from './calculation-result.effects';
