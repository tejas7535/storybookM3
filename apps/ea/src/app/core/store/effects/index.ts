export * from './calculation-parameters';
export * from './calculation-result';
