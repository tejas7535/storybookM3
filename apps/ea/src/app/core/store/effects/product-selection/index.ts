export * from './product-selection.effects';
