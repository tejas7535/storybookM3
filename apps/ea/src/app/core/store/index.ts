export * from './facades';
export * from './reducers';
export * from './selectors';
export * from './store.module';
