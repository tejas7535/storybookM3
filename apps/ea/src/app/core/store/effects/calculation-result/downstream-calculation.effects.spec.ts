/* eslint-disable jest/expect-expect */

import { HttpErrorResponse, provideHttpClient } from '@angular/common/http';
import { provideHttpClientTesting } from '@angular/common/http/testing';

import { of, ReplaySubject, throwError } from 'rxjs';

import { DownstreamCalculationService } from '@ea/core/services/downstream-calculation.service';
import { DownstreamCalculationInputsService } from '@ea/core/services/downstream-calculation-inputs.service';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';
import { Actions } from '@ngrx/effects';
import { provideMockActions } from '@ngrx/effects/testing';
import { provideMockStore } from '@ngrx/store/testing';
import { Context, marbles } from 'rxjs-marbles';

import { CO2DownstreamCalculationActions } from '../../actions';
import {
  CalculationParametersFacade,
  ProductSelectionFacade,
} from '../../facades';
import {
  CalculationParametersOperationConditions,
  LoadCaseData,
} from '../../models';
import { DownstreamCalculationEffects } from './downstream-calculation.effects';

jest.mock('@ea/shared/helper/downstream-error-helper', () => ({
  parseErrorObject: jest.fn(() => ['test error']),
}));

const downstreamCalculationServiceMock = {
  getDownstreamCalculation: jest.fn(),
};

const downstreamCalculationInputsServiceMock = {
  formatDownstreamInputs: jest.fn(),
};

describe('DownstreamCalculationEffects', () => {
  let spectator: SpectatorService<DownstreamCalculationEffects>;
  let actions$: any;
  let action: any;
  let effects: DownstreamCalculationEffects;

  const operatingConditionsWithType = (
    type: CalculationParametersOperationConditions['lubrication']['lubricationSelection']
  ) =>
    ({
      lubrication: {
        lubricationSelection: type,
        greaseType: 'grease',
        grease: {
          selection: 'typeOfGrease',
          typeOfGrease: {
            typeOfGrease: 'LB_FAG_MULTI_2',
          },
        },
        oilBath: {
          viscosity: {
            ny40: undefined,
            ny100: undefined,
          },
        },
        oilMist: {
          isoVgClass: {
            isoVgClass: 150,
          },
        },
        recirculatingOil: {
          selection: 'viscosity',
          viscosity: {
            ny40: 20,
            ny100: 50,
          },
          oilFlow: 20,
          oilTemperatureDifference: 50,
          externalHeatFlow: 45,
        },
      },
      energySource: {
        type: 'fossil',
        fossil: {
          fossilOrigin: 'LB_DIESEL_FOSSIL',
        },
      },
      time: 786,
      ambientTemperature: 50,
      conditionOfRotation: 'innerring',
    }) as unknown as Partial<CalculationParametersOperationConditions> as CalculationParametersOperationConditions;

  const loadcaseData: LoadCaseData = {
    load: {
      axialLoad: 20,
      radialLoad: 30,
    },
    rotation: {
      typeOfMotion: 'LB_ROTATING',
      rotationalSpeed: undefined,
      shiftFrequency: undefined,
      shiftAngle: undefined,
    },
    operatingTemperature: 70,
    operatingTime: 40,
    loadCaseName: 'Loadcase 1',
  };

  const operatingConditionsSubject$ =
    new ReplaySubject<CalculationParametersOperationConditions>();

  const loadcasesSubject$ = new ReplaySubject<
    { loadcaseName: string; index: number }[]
  >();

  const createEffectService = createServiceFactory({
    service: DownstreamCalculationEffects,
    providers: [
      provideHttpClient(),
      provideHttpClientTesting(),
      provideMockActions(() => actions$),
      provideMockStore({}),
      {
        provide: ProductSelectionFacade,
        useValue: {
          bearingDesignation$: of('bearing-123'),
        },
      },
      {
        provide: CalculationParametersFacade,
        useValue: {
          getLoadcases$: loadcasesSubject$.asObservable(),
          operationConditions$: operatingConditionsSubject$.asObservable(),
        },
      },
      {
        provide: DownstreamCalculationService,
        useValue: downstreamCalculationServiceMock,
      },
      {
        provide: DownstreamCalculationInputsService,
        useValue: downstreamCalculationInputsServiceMock,
      },
    ],
  });

  beforeEach(() => {
    spectator = createEffectService();
    actions$ = spectator.inject(Actions);

    effects = spectator.inject(DownstreamCalculationEffects);
    loadcasesSubject$.next([]);
  });

  describe('fetchDownstreamCalculation', () => {
    let fetchSpy: jest.SpyInstance;
    let formatInputSpy: jest.SpyInstance;

    const assertFetchDownstreamCalculation = (
      m: Context,
      a: ReturnType<
        typeof CO2DownstreamCalculationActions.fetchDownstreamCalculation
      >
    ) => {
      const expected = m.cold('-b', {
        b: CO2DownstreamCalculationActions.setDownstreamCalculationResult({
          result: 'result-from-service' as any,
          inputs: 'inputs-from-service' as any,
        }),
      });

      actions$ = m.hot('-a', { a });

      m.expect(effects.fetchDownstreamCalculation$).toBeObservable(expected);
      m.flush();

      expect(fetchSpy).toHaveBeenCalled();
      expect(fetchSpy.mock.calls).toMatchSnapshot();

      expect(formatInputSpy).toBeCalled();
    };

    beforeEach(() => {
      downstreamCalculationServiceMock.getDownstreamCalculation.mockReset();
      downstreamCalculationInputsServiceMock.formatDownstreamInputs.mockReset();

      fetchSpy = jest
        .spyOn(downstreamCalculationServiceMock, 'getDownstreamCalculation')
        .mockImplementation(() => of('result-from-service'));

      formatInputSpy = jest
        .spyOn(downstreamCalculationInputsServiceMock, 'formatDownstreamInputs')
        .mockImplementation(() => 'inputs-from-service');
    });

    afterEach(() => {
      fetchSpy.mockReset();
    });

    describe('when grease lubrication type is selected', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next(operatingConditionsWithType('grease'));
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when oilBath lubrication type is selected', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next(
          operatingConditionsWithType('oilBath')
        );
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when oilMist lubrication type is selected', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next(
          operatingConditionsWithType('oilMist')
        );
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when recirculatingOil lubrication type is selected', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next(
          operatingConditionsWithType('recirculatingOil')
        );
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when recirculatingOil lubrication type with isoVGClas and electic sources is selected', () => {
      beforeEach(() => {
        const base = operatingConditionsWithType('recirculatingOil');
        operatingConditionsSubject$.next({
          ...base,
          lubrication: {
            ...base.lubrication,
            recirculatingOil: {
              ...base.lubrication.recirculatingOil,
              selection: 'isoVgClass',
              isoVgClass: {
                isoVgClass: 50,
              },
              oilFlow: 20,
              oilTemperatureDifference: 30,
              externalHeatFlow: 15,
            },
          },
          energySource: {
            electric: {
              electricityRegion: 'LB_ARGENTINA',
            },
            type: 'electric',
          },
        });
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when single loadcase is provided', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next({
          ...operatingConditionsWithType('grease'),
          loadCaseData: [
            {
              ...loadcaseData,
            },
          ],
        });
        loadcasesSubject$.next([
          {
            loadcaseName: 'Cool loadcase 1',
            index: 0,
          },
        ]);
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when multiple loadcases are provided', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next({
          ...operatingConditionsWithType('grease'),
          lubrication: {
            ...operatingConditionsWithType('grease').lubrication,
            grease: {
              selection: 'isoVgClass',

              isoVgClass: {
                isoVgClass: 10,
              },
            } as Partial<
              CalculationParametersOperationConditions['lubrication']['grease']
            > as CalculationParametersOperationConditions['lubrication']['grease'],
          },
          loadCaseData: [
            {
              ...loadcaseData,
            },
            {
              ...loadcaseData,
              load: {
                axialLoad: undefined,
                radialLoad: undefined,
              },
            },
          ],
        });
        loadcasesSubject$.next([
          {
            loadcaseName: 'Cool loadcase 1',
            index: 0,
          },
          {
            loadcaseName: 'Cool loadcase 2',
            index: 1,
          },
        ]);
      });

      it('should fetch the downstream calculation', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when type is undefined', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next(
          operatingConditionsWithType(undefined)
        );
      });

      it('should handle gracefully', () =>
        marbles((m) => {
          action = CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          assertFetchDownstreamCalculation(m, action);
        })());
    });

    describe('when downstream service returns error', () => {
      beforeEach(() => {
        operatingConditionsSubject$.next({
          ...operatingConditionsWithType('grease'),
        });
      });

      it(
        'should dispatch setCalculationFailure action on error',
        marbles((m) => {
          const errorResponse = new HttpErrorResponse({
            error: { title: 'Test Error' },
            status: 500,
          });

          downstreamCalculationServiceMock.getDownstreamCalculation.mockReturnValueOnce(
            throwError(() => errorResponse)
          );

          const fetchAction =
            CO2DownstreamCalculationActions.fetchDownstreamCalculation();
          const completion =
            CO2DownstreamCalculationActions.setCalculationFailure({
              errors: ['test error'],
            });

          actions$ = m.hot('-a', { a: fetchAction });
          const expected = m.cold('-b', { b: completion });

          m.expect(effects.fetchDownstreamCalculation$).toBeObservable(
            expected
          );
        })
      );
    });
  });
});
