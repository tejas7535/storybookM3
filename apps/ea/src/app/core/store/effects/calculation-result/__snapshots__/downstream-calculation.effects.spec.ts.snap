// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when grease lubrication type is selected should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": undefined,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": "LB_FAG_MULTI_2",
        "isoVgClass": undefined,
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_GREASE_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": undefined,
        "oilTemperatureRise": undefined,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": "LB_ARCANOL_GREASE",
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when multiple loadcases are provided should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [
        Object {
          "axialLoad": 20,
          "designation": "",
          "movementType": "LB_ROTATING",
          "operatingTemperature": 70,
          "radialLoad": 30,
          "speed": 0,
          "timePortion": 40,
        },
        Object {
          "axialLoad": 0,
          "designation": "",
          "movementType": "LB_ROTATING",
          "operatingTemperature": 70,
          "radialLoad": 0,
          "speed": 0,
          "timePortion": 40,
        },
      ],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": undefined,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": undefined,
        "isoVgClass": "LB_ISO_VG_10",
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_GREASE_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": undefined,
        "oilTemperatureRise": undefined,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": "LB_ISO_VG_CLASS",
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when oilBath lubrication type is selected should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": undefined,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": undefined,
        "isoVgClass": undefined,
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_OIL_BATH_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": undefined,
        "oilTemperatureRise": undefined,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": undefined,
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when oilMist lubrication type is selected should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": undefined,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": undefined,
        "isoVgClass": undefined,
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_OIL_MIST_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": undefined,
        "oilTemperatureRise": undefined,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": undefined,
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when recirculatingOil lubrication type is selected should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": 45,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": undefined,
        "isoVgClass": undefined,
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_RECIRCULATING_OIL_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": 20,
        "oilTemperatureRise": 50,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": "LB_ENTER_VISCOSITIES",
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when recirculatingOil lubrication type with isoVGClas and electic sources is selected should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [],
      "operatingConditions": Object {
        "electricEmissionFactor": "LB_ARGENTINA",
        "emissionFactor": "LB_ELECTRIC_ENERGY",
        "externalHeatFlow": 15,
        "fossilEmissionFactor": undefined,
        "greaseType": undefined,
        "isoVgClass": "LB_ISO_VG_50",
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_RECIRCULATING_OIL_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": 20,
        "oilTemperatureRise": 30,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": "LB_ISO_VG_CLASS",
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when single loadcase is provided should fetch the downstream calculation 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [
        Object {
          "axialLoad": 20,
          "designation": "",
          "movementType": "LB_ROTATING",
          "operatingTemperature": 70,
          "radialLoad": 30,
          "speed": 0,
          "timePortion": 100,
        },
      ],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": undefined,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": "LB_FAG_MULTI_2",
        "isoVgClass": undefined,
        "isoVgClassCalculated": undefined,
        "lubricationMethod": "LB_GREASE_LUBRICATION",
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": undefined,
        "oilTemperatureRise": undefined,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": "LB_ARCANOL_GREASE",
      },
    },
  ],
]
`;

exports[`DownstreamCalculationEffects fetchDownstreamCalculation when type is undefined should handle gracefully 1`] = `
Array [
  Array [
    "bearing-123",
    Object {
      "loadcases": Array [],
      "operatingConditions": Object {
        "electricEmissionFactor": undefined,
        "emissionFactor": "LB_FOSSIL_ENERGY",
        "externalHeatFlow": undefined,
        "fossilEmissionFactor": "LB_DIESEL_FOSSIL",
        "greaseType": undefined,
        "isoVgClass": undefined,
        "isoVgClassCalculated": undefined,
        "lubricationMethod": undefined,
        "ny100": undefined,
        "ny40": undefined,
        "oilFlow": undefined,
        "oilTemperatureRise": undefined,
        "operatingTimeInHours": 786,
        "rotationType": "LB_ROTATING_INNERRING",
        "temperature": 50,
        "viscosityDefinition": undefined,
      },
    },
  ],
]
`;
